### DISCUSSION TITLE
Cache IP in EFI ACPI volume when chainloading images

### DISCUSSION THREAD

**Discussion by omniproc** [ipxe/ipxe#1369] [General]
I'm not entirely sure if this is due to the custom build I'm using but so far I've not been able to get this working. So, apperently [some EFI bootloaders](https://github.com/vmware/esx-boot/blob/95d864a981aaa256b069ed669519001f91723e68/uefi/efiutils/volume.c#L210) check the ACPI PciRoot volume for IP configurations. 

I'm first booting a iPXE image that runs a iPXE script. In that process, iPXE requests a IP via DHCP using the `dhcp` command. After that I'm chainloading the next bootloader using `chain https://path/to/bootx64.efi`. However the PciRoot looks something like this `PciRoot(0x3)/Pci(0x1,0x0)/Pci(0x0,0x0)/MAC(D06726D151E9,0x1)`. So, the IP configuration that iPXE had get's lost in the process. 

It seems that everything still works. The chainloaded bootloader is able to perform the next steps of downloading additional modules (to some degree), so the IP configuration that iPXE requested seems to perist. I'm just not entirely sure how that's done if not via the ACPI volume and if it might be a good idea to configure it there too.

Any insights from someone more familiar with the EFI chaining process would be great.

**Comment by NiKiZe** [comment]
Since this is a question, rather than a identified bug I'm moving it from issue to discussion.

Many loaders don't care about IP config, they just call the already registered protocols to do the work, such as "grab me this tftp url" or similar.

**Comment by omniproc** [comment]
Understood. So, it wouldn't hurt to add the information there in case a loader decides it wants to do something with it?

**Comment by mcb30** [comment]
UEFI is somewhat haphazard in how information is conveyed between stages of the boot process.

For standard PXE (TFTP) boot, UEFI uses a modified version of the old BIOS PXE specification, with an `EFI_PXE_BASE_CODE_PROTOCOL` that exposes the various DHCP-like packets received by the previous stage.  Subsequent stages are expected to check for the existence of this protocol and query it to extract the relevant information.

For HTTP boot, UEFI exposes a small subset of the information in the device path (which is what you are calling the "EFI ACPI volume").  Large chunks of required information are simply not exposed at all.

Even when UEFI firmware has an IPv4 address, you'll often see device paths that contain `0.0.0.0` as the IP address, since this seems to be the de facto standard used when the IP address was obtained via DHCP.

iPXE chooses not to put IPv4 or IPv6 configuration in the device path, on the basis that:
* No consumer can rely on it anyway, since consumers will have to use `EFI_PXE_BASE_CODE_PROTOCOL` for compatibility with other boot firmware
* UEFI has a very limited concept of the world, in which the user is forced to choose IPv4 or IPv6 before attempting the boot, and can use only DHCPv6 for IPv6. iPXE has a more general model than UEFI: for example, you can use SLAAC (rather than DHCPv6) to obtain an IPv6 address, simultaneously use DHCPv4 to obtain an IPv4 address, then boot from a URL containing a DNS name that has both AAAA (IPv6) and A (IPv4) records.  In this circumstance, there wouldn't be any sensible device path that could be created to represent the IPv4/IPv6 configuration.

Absent some very good reason to include the configuration (such as a newer version of the UEFI spec finally bothering to specify what it should be), I'm not going to change this since it would be likely to break something somewhere.

**Reply by omniproc** [reply]
Thanks for the explanation.
